sklearn/benchmark.py

```diff
@@ -0,0 +1,188 @@
+    y : array-like, optional
+    fit_params : dict, optional
+    vary_n_samples : bool, default=True
+        Whether to benchmark for various random sample sizes.
+    n_fits : int, default=100
+        Maximum number of fits to make while benchmarking.
+    time_budget : int, default=300
+        Maximum number of seconds to use overall.  Current fit will
+        be stopped if the budget is exceeded.
+    profile_memory : bool, default=True
+        Whether to include memory (or just time) profiling. Memory
+        profiling will slow down fitting, and hence make fit_time
+        estimates more approximate.
+
+    Returns
+    -------
+    results : dict
+        The following keys are each mapped to an array:
+
+        n_samples
+            The number of samples
+        fit_time
+            In seconds
+        peak_memory
+            The memory used at peak of fitting, in KiB.
+        model_memory
+            The memory in use at the end of fitting, minus that at the
+            beginning, in KiB.
+
+    models : dict
+        keys 'peak_memory', 'model_memory' and 'fit_time' map to polynomial
+        GP regressors whose input is n_samples  and whose
+        outputs are each of those targets.
+
+    errors : list of dicts
+        lists the parameters that resulted in exceptions
+
+    """
+    fit_times = []
+    length = _num_samples(X)
+    estimator = est
+    parameters = est_params
+    n_samples = []
+    model_memory = []
+    SAMPLES = 8
+
+    if profile_memory:
+        try:
+            import psutil
+        except ImportError:
+            # To-Do : Make error message more informative
+            raise ImportError("Please install psutil")
+
+    if parameters is not None:
+        estimator.set_params(**parameters)
+
+    if not vary_n_samples:
+        for _ in range(0, n_fits):
+            if y is None:
+                start_time = time.time()
+                if fit_params is not None:
+                    estimator.fit(X, **fit_params)
+                else:
+                    estimator.fit(X)
+                time_taken = time.time() - start_time
+                fit_times.append(time_taken)
+                n_samples.append(length)
+            else:
+                start_time = time.time()
+                if fit_params is not None:
+                    estimator.fit(X, y, **fit_params)
+                else:
+                    estimator.fit(X, y)
+                time_taken = time.time() - start_time
+                fit_times.append(time_taken)
+
+    if vary_n_samples:
+        for _ in range(0, n_fits):
+            if SAMPLES < length:
+                SAMPLES *= 2
+            else:
+                SAMPLES = length
+            n_samples.append(SAMPLES)
+            train_size = n_samples[_] - 1
+            X_vary = X[:train_size]
+            y_vary = y[:train_size]
+            if y is None:
+                if fit_params is not None:
+                    start_time = time.time()
```

Comment:
please do not repeat yourself. Use a helper function, or set `fit_params={}` when it it is otherwise None. Similarly, you can use `*args` instead of passing `y` in directly and avoid repetition.